src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -152,13 +152,14 @@ internal ItemManager(RecentItemsManager recentItemsManager)
 
             var filterReason = Helpers.GetFilterReason(data.Trigger);
 
-            // If the session was created/maintained out of Roslyn, e.g. in debugger; no properties are set and we should use data.Snapshot.
-            // However, we prefer using the original snapshot in some projection scenarios.
-            var snapshotForDocument = Helpers.TryGetInitialTriggerLocation(session, out var triggerLocation)
-                ? triggerLocation.Snapshot
-                : data.Snapshot;
-
-            var document = snapshotForDocument.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+            // We prefer using the original snapshot, which should always be available from items provided by Roslyn CompletionSource.
+            // Just to be defensive, if it's not found then use data.Snapshot instead.
+            var snapshotForDocument = data.InitialSortedList.FirstOrDefault(item => item.Properties.ContainsProperty(CompletionSource.TriggerLocation)) is VSCompletionItem firstItem &&
+                                        firstItem.Properties.GetProperty(CompletionSource.TriggerLocation) is SnapshotPoint triggerLocation
+                                        ? triggerLocation.Snapshot
+                                        : data.Snapshot;
```

Comment:
Data is captured here in the lambda, most of the other lambdas are static in the method here.

Reply:
this is just really hard to read.  can you make this a simple helper function.

Reply:
> which should always be available

If it shoudl always be available, we should either assert/contract so we get a stack when something breaks expectations.  being defensive just means unexplainable bugs happen and we have no idea what is going on.

Reply:
> Data is captured here in the lambda, most of the other lambdas are static in the method here.

I don't see anything got captured, but I did extract a local function here

> If it shoudl always be available, we should either assert/contract so we get a stack when something breaks expectations. being defensive just means unexplainable bugs happen and we have no idea what is going on.

Add some comments to clearify